hyprtester/plugin/src/main.cpp

```diff
@@ -48,11 +49,85 @@ static SDispatchResult snapMove(std::string in) {
     return {};
 }
 
+class CTestKeyboard : public IKeyboard {
+  public:
+    static SP<CTestKeyboard> create(bool isVirtual) {
+        auto keeb            = SP<CTestKeyboard>(new CTestKeyboard());
+        keeb->m_self         = keeb;
+        keeb->m_isVirtual    = isVirtual;
+        keeb->m_share_states = !isVirtual;
+        return keeb;
+    }
+
+    virtual bool isVirtual() {
+        return m_isVirtual;
+    }
+
+    virtual SP<Aquamarine::IKeyboard> aq() {
+        return nullptr;
+    }
+
+    void sendKey(uint32_t key, bool pressed) {
+        auto event = IKeyboard::SKeyEvent{
+            .timeMs  = static_cast<uint32_t>(Time::millis(Time::steadyNow())),
+            .keycode = key,
+            .state   = pressed ? WL_KEYBOARD_KEY_STATE_PRESSED : WL_KEYBOARD_KEY_STATE_RELEASED,
+        };
+        updatePressed(event.keycode, pressed);
+        m_keyboardEvents.key.emit(event);
+    }
+
+  private:
+    bool m_isVirtual;
+};
+
+static SDispatchResult vkb(std::string in) {
+    auto tkb0 = CTestKeyboard::create(false);
+    auto tkb1 = CTestKeyboard::create(false);
+    auto vkb0 = CTestKeyboard::create(true);
+
+    g_pInputManager->newKeyboard(tkb0);
+    g_pInputManager->newKeyboard(tkb1);
+    g_pInputManager->newKeyboard(vkb0);
+
+    const auto&    PRESSED = g_pInputManager->getKeysFromAllKBs();
+    const uint32_t TESTKEY = 1;
+
+    tkb0->sendKey(TESTKEY, true);
+    if (!std::ranges::contains(PRESSED, TESTKEY)) {
+        return {
+            .success = false,
+            .error   = "Expected pressed key not found",
+        };
+    }
+
+    tkb1->sendKey(TESTKEY, true);
+    tkb0->sendKey(TESTKEY, false);
+    if (!std::ranges::contains(PRESSED, TESTKEY)) {
+        return {
+            .success = false,
+            .error   = "Expected pressed key not found (kb share state)",
+        };
+    }
+
+    vkb0->sendKey(TESTKEY, true);
+    tkb1->sendKey(TESTKEY, false);
+    if (std::ranges::contains(PRESSED, TESTKEY)) {
+        return {
+            .success = false,
+            .error   = "Expected released key found in pressed (vkb no share state)",
+        };
+    }
+
+    return {};
```

Comment:
if you could remove the keyboards here after the tests are done to end on a clean state